include/mqtt/endpoint.hpp

```diff
@@ -7431,11 +7441,22 @@ class endpoint : public std::enable_shared_from_this<endpoint<Socket, Mutex, Loc
         auto fixed_header = static_cast<std::uint8_t>(*b);
         switch (get_control_packet_type(fixed_header)) {
         case control_packet_type::publish: {
-            auto sp = std::make_shared<std::string>(b, e);
-            restore_serialized_message(basic_publish_message<PacketIdBytes>(sp->begin(), sp->end()), sp);
+            auto size = static_cast<std::size_t>(std::distance(b, e));
+            shared_ptr_array spa { new char[size] };
+            std::copy(b, e, &spa[0]);
+            restore_serialized_message(
+                basic_publish_message<PacketIdBytes>(
+                    buffer(spa.get(), size)
+                ),
+                spa
+            );
         } break;
         case control_packet_type::pubrel: {
-            restore_serialized_message(basic_pubrel_message<PacketIdBytes>(b, e));
+            restore_serialized_message(
+                basic_pubrel_message<PacketIdBytes>(
+                    buffer(&*b, static_cast<std::size_t>(std::distance(b, e)))
```

Comment:
This assumes that the provided iterator is a random-access iterator (or, at the very least, is contiguous)

The iterator may potentially be forward only

Probably should use a static_assert that the iterator is random-access.

Reply:
Fixed.
I insert static_assert just before `restore_serialized_message(` but the iterator category is always checked. I think that it is acceptable.

Reply:
I considered again.

The current code is as follows:

```cpp
    template <typename Iterator>
    typename std::enable_if<std::is_convertible<typename Iterator::value_type, char>::value>::type
    restore_serialized_message(packet_id_t /*packet_id*/, Iterator b, Iterator e) {
        if (b == e) return;

        auto fixed_header = static_cast<std::uint8_t>(*b);
        switch (get_control_packet_type(fixed_header)) {
        case control_packet_type::publish: {
            auto size = static_cast<std::size_t>(std::distance(b, e));
            shared_ptr_array spa { new char[size] };
            std::copy(b, e, spa.get());
            restore_serialized_message(
                basic_publish_message<PacketIdBytes>(
                    buffer(spa.get(), size)
                ),
                spa
            );
        } break;
        case control_packet_type::pubrel: {
            static_assert(
                std::is_same<
                    typename std::iterator_traits<Iterator>::iterator_category,
                    std::random_access_iterator_tag
                >::value
            );
            restore_serialized_message(
                basic_pubrel_message<PacketIdBytes>(
                    buffer(&*b, static_cast<std::size_t>(std::distance(b, e)))
                )
            );
        } break;
        default:
            throw protocol_error();
            break;
        }
    }
```

static_assert is confusing.

Should be as follows ?

```cpp
    template <typename Iterator>
    typename std::enable_if<std::is_convertible<typename Iterator::value_type, char>::value>::type
    restore_serialized_message(packet_id_t /*packet_id*/, Iterator b, Iterator e) {
        static_assert(
            std::is_same<
                typename std::iterator_traits<Iterator>::iterator_category,
                std::random_access_iterator_tag
            >::value
        );
        if (b == e) return;

        auto fixed_header = static_cast<std::uint8_t>(*b);
        switch (get_control_packet_type(fixed_header)) {
        case control_packet_type::publish: {
            auto size = static_cast<std::size_t>(std::distance(b, e));
            shared_ptr_array spa { new char[size] };
            std::copy(b, e, spa.get());
            restore_serialized_message(
                basic_publish_message<PacketIdBytes>(
                    buffer(spa.get(), size)
                ),
                spa
            );
        } break;
        case control_packet_type::pubrel: {
            restore_serialized_message(
                basic_pubrel_message<PacketIdBytes>(
                    buffer(&*b, static_cast<std::size_t>(std::distance(b, e)))
                )
            );
        } break;
        default:
            throw protocol_error();
            break;
        }
    }
```

Reply:
I think the second one is better.

Reply:
Fixed.